src/hope/apps/periodic_data_update/notifications.py

```diff
@@ -0,0 +1,146 @@
+        self.pdu_creator = self.pdu_online_edit.created_by
+        self.pdu_creation_date = self.pdu_online_edit.created_at
+        self.email_subject = self.ACTION_PREPARE_EMAIL_BODIES_MAP[self.action]["subject"]
+        self.action_name = self.ACTION_PREPARE_EMAIL_BODIES_MAP[self.action]["action_name"]
+        self.recipient_title = self.ACTION_PREPARE_EMAIL_BODIES_MAP[self.action]["recipient_title"]
+        self.user_recipients = self._prepare_user_recipients()
+        self.email = self._prepare_email()
+        self.enable_email_notification = self.pdu_online_edit.business_area.enable_email_notification
+
+    def _prepare_user_recipients(self) -> QuerySet[User]:
+        if self.action == self.ACTION_SEND_BACK:
+            # For send back action, notify the creator
+            return User.objects.filter(id=self.pdu_creator.id).exclude(id=self.action_user.id)
+
+        else:
+            permission = self.ACTION_TO_RECIPIENTS_PERMISSIONS_MAP[self.action]
+            business_area = self.pdu_online_edit.business_area
+            program = self.pdu_online_edit.program
+
+            # Get authorized users for this PDU Edit
+            authorized_user_ids = list(self.pdu_online_edit.authorized_users.values_list('id', flat=True))
+
+            if not authorized_user_ids:
+                return User.objects.none()
+
+            role_assignments = (
+                RoleAssignment.objects.filter(
+                    Q(role__permissions__contains=[permission])
+                    & Q(business_area=business_area)
+                    & (Q(program=None) | Q(program=program))
+                )
+                .exclude(expiry_date__lt=timezone.now())
+                .distinct()
+            )
+
+            users = (
+                User.objects.filter(
+                    Q(role_assignments__in=role_assignments) | Q(partner__role_assignments__in=role_assignments)
+                )
+                .filter(id__in=authorized_user_ids)  # Only authorized users
+                .exclude(id=self.action_user.id)
+                .distinct()
+            )
+
+            if settings.ENV == "prod":
+                users = users.exclude(is_superuser=True)
+            return users
+
+    def _prepare_email(self) -> MailjetClient:
+        body_variables = self._prepare_body_variables()
+        return MailjetClient(
+            mailjet_template_id=config.MAILJET_TEMPLATE_PDU_ONLINE_EDIT_NOTIFICATION,
+            subject=self.email_subject,
+            recipients=[user_recipient.email for user_recipient in self.user_recipients],
+            ccs=[self.action_user.email],
+            variables=body_variables,
+        )
+
+    def send_email_notification(self) -> None:
+        if config.SEND_PDU_ONLINE_EDIT_NOTIFICATION and self.enable_email_notification:
+            try:
+                self.email.send_email()
+            except Exception as e:  # pragma: no cover
+                logger.exception(e)
```

Comment:
logger.exception() automatically includes traceback so logger.exception("message") not logger.exception(e)